include/libvfio-user.h

```diff
@@ -268,33 +246,106 @@ typedef enum {
+// FIXME: what else do we need? multi-function flag in "header type" ? rid?
+
+/**
+ * Add a PCI capability to PCI config space.
+ *
+ * Certain standard capabilities are handled entirely within the library:
+ *
+ * PCI_CAP_ID_EXP (pxcap)
+ * PCI_CAP_ID_MSIX (msixcap)
+ * PCI_CAP_ID_PM (pmcap)
+ *
+ * However, they must still be explicitly initialized and added here.
  *
- * @returns 0 on success, -1 on failure and sets errno.
+ * FIXME: are we really making the user init these too? why? Is there a better
+ * API?
+ *
+ * The contents of @data is copied in. It must start with either a struct
+ * pci_cap or a struct pci_extcap, with the ID field set (the 'next' field is
+ * ignored).  If the capability is not one of the handled ones above (e.g.
+ * PCI_CAP_ID_VNDR), the library will allow all reads and disallow all writes
+ * for the capability from the client.
+ *
+ * Note the size of the capability is derived either from the ID, or from an
+ * embedded length field for PCI_(EXT_)CAP_ID_VNDR.
+ *
+ * @vfu_ctx: the libvfio-user context
+ * @cap: capability ID to add
+ * @extended: true if an extended capability (PCI_EXT_CAP_ID_*)
+ * @data: capability data, not including the capability header
  */
 int
-vfu_pci_setup_config_hdr(vfu_ctx_t *vfu_ctx, vfu_pci_hdr_id_t id,
-                         vfu_pci_hdr_ss_t ss, vfu_pci_hdr_cc_t cc,
-                         vfu_pci_type_t pci_type,
-                         int revision __attribute__((unused)));
+vfu_pci_add_capability(vfu_ctx_t *vfu_ctx, uint16_t cap_id, bool extended,
+                       void *data);
+
+/**
+ * Callback function that gets called when a capability is accessed.
+ *
+ * @pvt: private pointer
+ * @id: capability ID being accessed
+ * @buf: pointer to data being read or written
+ * @count: number of bytes being read or written
+ * @offset: offset within the capability (including the header)
+ * @is_write: whether the capability is read or written
+ *
+ * @returns the number of bytes read or written
+ *
+ * FIXME: cb would have to identify each cap by "pvt" - is that OK?
+ * FIXME: this can already be handled by region callback - do we need this
+ * too?
+ */
+typedef ssize_t (vfu_cap_access_cb_t) (void *pvt, uint8_t id,
+                                       char *buf, size_t count,
+                                       loff_t offset, bool is_write);
 
-//TODO: Support variable size capabilities.
 /**
- * Setup PCI capabilities.
+ * Add a PCI capability to PCI config space with a callback.
+ *
+ * For capabilities not handled by the library, (in particular PCI_CAP_ID_VNDR
+ * and PCI_EXT_CAP_ID_VNDR), this can be used to add a capability where reads
+ * and writes are delegated to a callback.
+ *
+ * Note the size of the capability is derived either from the ID, or from an
+ * embedded length field for PCI_(EXT_)CAP_ID_VNDR.
+ *
  * @vfu_ctx: the libvfio-user context
- * @caps: array of (vfu_cap_t *)
- * *nr_caps: number of elements in @caps
+ * @cap: capability ID to add
```

Comment:
I would add "(PCI_CAP_ID_ or PCI_EXT_CAP_ID_ in <linux/pci_regs.h>)".

Reply:
good idea